domain/modelconfig/state/controllerstate.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	coredatabase "github.com/juju/juju/core/database"
+	coremodel "github.com/juju/juju/core/model"
+	"github.com/juju/juju/domain"
+	modelerrors "github.com/juju/juju/domain/model/errors"
+	"github.com/juju/juju/domain/secretbackend"
+	backenderrors "github.com/juju/juju/domain/secretbackend/errors"
+	"github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/secrets/provider"
+	"github.com/juju/juju/internal/secrets/provider/kubernetes"
+)
+
+// ControllerState is a reference to the underlying data accessor for data
+// in the controller database.
+type ControllerState struct {
+	*domain.StateBase
+}
+
+// NewControllerState creates a new state struct for querying the controller state.
+func NewControllerState(factory coredatabase.TxnRunnerFactory) *ControllerState {
+	return &ControllerState{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetModelSecretBackend returns the secret backend name for the specified model.
+func (s *ControllerState) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, error) {
+	db, err := s.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	var backendName string
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		modelBackendDetails, err := s.getModelSecretBackendDetails(ctx, modelUUID, tx)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		backendName = modelBackendDetails.SecretBackendName
+		switch modelBackendDetails.Type {
+		case coremodel.IAAS:
+			if backendName == provider.Internal {
+				backendName = provider.Auto
+			}
+		case coremodel.CAAS:
+			if backendName == kubernetes.BackendName {
+				backendName = provider.Auto
+			}
```

Comment:
This logic belongs in the service. The role of state is just to hydrate what's in the db. Business logic to interpret the data should be in the service.